### ISSUE TITLE
Many Cg errors inside log

### ISSUE THREAD

**Issue by kPYKx7Ddw4n1aIKZ** [OGRECave/scape#6]
#### System Information
- Ogre Version: v1.11.0
- Operating System / Platform: Debian GNU/Linux 9 (stretch)
- RenderSystem: RenderSystem_GL
- Graphics Device: Nvidia GeForce GTX 660

#### Detailled description
I managed to compile the tool.
It does start and I can see the window.

It displays the SkyBox and the terrain (in which im not sure, if it does that right).

Most importantly, there are several errors in the log, saying that the GPU operation shaders didn't compile.
Mainly they are issues complaining about some "exceeded limits".

I don't think this is an issue of my hardware, since I use a relatively up to date graphics card.
I think this has something to do with the profiles in the GPU2DOperation.material file.

#### Scape.log (only the important part)
```
error C6002: Instruction limit of 1024 exceeded; 1296 instructions needed to compile program
error C6002: Instruction limit of 1024 exceeded; 1320 instructions needed to compile program
error C6007: Constant register limit exceeded; more than 32 constant registers needed to compiled program
```

**Comment by kPYKx7Ddw4n1aIKZ** [comment]
I think I have found a potential solution to this.

One would have to replace the Cg profiles in lines like this:
https://github.com/OGRECave/scape/blob/6708cf212b7c8b6e1ff83bc438743c14c59acec3/Dependencies/EditorMedia/Materials/Scripts/GPU2DOperation.material#L14

For me, it worked when I replaced the `arbvp1` by `glslv` and `arbfp1` by `glslf`.
(At least no errors occured.)

The drawback would be that only GPUs supporting OpenGL 2.0 and newer would be able to run this.
Someone else would have to decide if that would be an issue.
[(Source)](https://developer.nvidia.com/cg-profiles)

I could provide a pull request for this, but i think it might be better to wait until #2 is solved.
(Especially line-endings and format.)

**Comment by AllianceDarkSylinc** [comment]
Hi!
I'm not the maintainer, but that is the correct solution.

The shader is too expensive and thus when compiled it surpasses what **_very_** old HW could do, hence arbvp1 fails to compile. You could say the shader was never compatible with pre-2.0 GL hardware anyway (unless someone somehow rewrites the shader in a way that Cg compiles it into short-enough ops that arb*p1 requirements are satisfied, if that's even possible)

Other profiles that can work are vp4 and fp4 which are NV-only profiles that are analogous to D3D's vs_3_0 and ps_3_0 profiles. Though, these profiles won't work on AMD and Intel cards.

Cheers

**Comment by darksylinc** [comment]
Hi!
I'm not the maintainer, but that is the correct solution.

The shader is too expensive and thus when compiled it surpasses what **_very_** old HW could do, hence arbvp1 fails to compile. You could say the shader was never compatible with pre-2.0 GL hardware anyway (unless someone somehow rewrites the shader in a way that Cg compiles it into short-enough ops that arb*p1 requirements are satisfied, if that's even possible)

Other profiles that can work are vp4 and fp4 which are NV-only profiles that are analogous to D3D's vs_3_0 and ps_3_0 profiles. Though, these profiles won't work on AMD and Intel cards.

Cheers

**Comment by paroj** [comment]
> The drawback would be that only GPUs supporting OpenGL 2.0 and newer would be able to run this.

the profiles are checked in order of declaration, so you can write to allow falling back to arbvp1
```
profiles vs_3_0 glslv arbvp1
```

however if the shader indeed requires vs_3_0 we should just drop arbvp1 as it can not possibly work.

**Comment by darksylinc** [comment]
arb*p1 should be dropped: In my experience, leaving it only encourages new bug reports from people with older cards who think their system should be supported (because we try to compile w/ arbvp1) thus leading them to believe the program accidentally broke support at some point and can just be easily fixed, which is not true.

**Comment by kPYKx7Ddw4n1aIKZ** [comment]
I tried to use glsl* everywhere.
The shaders compiled, but they didn't work as expected.
(The terrain wasn't visible. 😕)

Somehow it worked fine for me, when I used the gp4*p profile.
(The terrain was visible and all the GPU operations worked.)

Maybe I overlooked something?

**Comment by paroj** [comment]
the glsl* profiles translation is quite hacky on OGRE, to say the least:

**Comment by kPYKx7Ddw4n1aIKZ** [comment]
I looked through the OGRE examples and noticed that not even a single sample uses the glsl* profile.
So I guess this haven't been tested extensively.
But on the other hand, I guess that this isn't a problem of this program but OGRE itself?

A possible workaround would be, to just support gp4*p (and possibly *p40) on non DirectX environments until this gets fixed.

**Comment by paroj** [comment]
yes, I think we should make profile lines look like:
`profiles vs_3_0 vp30 glslv`

so the shaders will still work on NVidia/ GLSL while having the correct (albeit currently broken) fallback.

> I looked through the OGRE examples and noticed that not even a single sample uses the glsl* profile.

I have updated some DQ Shadowmapping shaders in OGRE core to glsl* (as they were GL2+ only anyway), but those have worked so far..

**Comment by kPYKx7Ddw4n1aIKZ** [comment]
> yes, I think we should make profile lines look like:
> `profiles vs_3_0 vp30 glslv`

I just tested the *p30 profile and I got again errors like:
`(0) : error C6002: Instruction limit of 1024 exceeded; 1131 instructions needed to compile program`

Using the *p40 profile however worked.

So I guess that the *p30 profile might still be too weak.

**Comment by paroj** [comment]
fixed by #10
